### PR TITLE
Metadata

### PR DESCRIPTION
**Pull request by achunBOX**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @achunBOX has signed the CLA. Thanks for the pull request!

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
+//
+//  BOXContentClient+Metadata.h
+//  BoxContentSDK
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h:3]: Sanitize all headers and implementation files based on existing source files.

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+ * Generate a request to create metadata information for a file.
+ *
+ * **NOTE** Defaults the scope property to @see BOXAPIScopeEnterprise.
+ *
+ * @param fileID File ID.
+ * @param template Template for metadata in file.
+ *
+ * @return A request that can be customized and then executed.
+ */
+- (BOXMetadataCreateRequest *)metadataCreateRequestWithFileID:(NSString *)fileID template:(NSString *)template info:(NSArray *)info;
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h]: Clarification for info. Aka. better name.

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ * @param fileID File ID.
+ * @param scope Scope to see which templates are available.
+ * @param template Template for metadata in file.
+ *
+ * @return A request that can be customized and then executed.
+ */
+- (BOXMetadataDeleteRequest *)metadataDeleteRequestWithFileID:(NSString *) fileID scope:(NSString *)scope template:(NSString *)template;
+
+/**
+ * Generate a request to create metadata information for a file.
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h:82]: State that all objects in info must be of type BOXMetadataTask.

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * Generate a request to update metadata information about a file.
+ *
+ * @param fileID File ID.
+ * @param scope Scope to see which templates are available.
+ * @param template Template for metadata in file.
+ *
+ * @return A request that can be customized and then executed.
+ */
+- (BOXMetadataUpdateRequest *)metadataUpdateRequestWithFileID:(NSString *)fileID scope:(NSString *)scope template:(NSString *)template updateInfo:(NSArray *)updateInfo;
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h]: Same thing as create.

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadata.m [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+#import "BOXMetadata.h"
+
+@implementation BOXMetadata
+
+- (instancetype)initWithJSON:(NSDictionary *)JSONData
+{
+    if (self = [super initWithJSON:JSONData]) {
+        self.JSONData = JSONData;
+        
+        self.modelID = [NSJSONSerialization box_ensureObjectForKey:@"$id"
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadata.m]: Make $id and the other constants.

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadata.m [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        self.template = [NSJSONSerialization box_ensureObjectForKey:@"$template"
+                                                       inDictionary:JSONData
+                                                    hasExpectedType:[NSString class]
+                                                        nullAllowed:NO];
+        
+        self.parent = [NSJSONSerialization box_ensureObjectForKey:@"$parent"
+                                                     inDictionary:JSONData
+                                                  hasExpectedType:[NSString class]
+                                                      nullAllowed:NO];
+        self.type = [NSJSONSerialization box_ensureObjectForKey:@"$type"
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadata.m]: Remove. Redundant.

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadata.m [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                                                   inDictionary:JSONData
+                                                hasExpectedType:[NSString class]
+                                                    nullAllowed:NO];
+        
+        NSMutableDictionary *info = [[NSMutableDictionary alloc]init];
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadata.m:46]: Comments to explain why this is here.

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadataTemplate.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
+//
+//  BOXMetadataTemplate.h
+//  BoxContentSDK
+//
+//  Created by Andrew Chun on 6/15/15.
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadataTemplate.h]: Explore advanced metadata search.

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadataTemplate.m [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        self.scope = [NSJSONSerialization box_ensureObjectForKey:@"scope"
+                                                    inDictionary:JSONData
+                                                 hasExpectedType:[NSString class]
+                                                     nullAllowed:NO];
+        self.displayName = [NSJSONSerialization box_ensureObjectForKey:@"displayName"
+                                                          inDictionary:JSONData
+                                                       hasExpectedType:[NSString class]
+                                                           nullAllowed:NO];
+        NSMutableArray *fields = [[NSMutableArray alloc]init];
+        for (NSDictionary *field in [NSJSONSerialization box_ensureObjectForKey:@"fields"
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadataTemplate.m]: Put NSJSONSerialization box_ensure... in a separate line for clarity.

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadataUpdateTask.h [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * Remove Operation.
+     */
+    BOXMetadataUpdateREMOVE,
+    
+    /**
+     * Test Operation.
+     */
+    BOXMetadataUpdateTEST
+} BOXMetadataUpdateOps;
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadataUpdateTask.h]: Rename to BOXMetadataUpdateOperation

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadataUpdateTask.m [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+#import "BOXLog.h"
+
+@implementation BOXMetadataUpdateTask
+
+- (instancetype)initWithOperation:(BOXMetadataUpdateOps)operation path:(NSString *)path value:(NSString *)value
+{
+    return [self initWithOperation:operation path:path value:value validation:YES];
+}
+
+- (instancetype)initWithOperation:(BOXMetadataUpdateOps)operation path:(NSString *)path value:(NSString *)value validation:(BOOL)validate
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadataUpdateTask.m]: Try to clean up the validation portion of this initializer.

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadataUpdateTask.m [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+- (void)validate
+{
+    BOXAssert([self BOXMetadataUpdateOpsToString:self.operation], @"The operation parameter for BOXMetadataUpdateTask cannot be nil. Please make sure to use an ENUM within BOXMetadataUpdateOps.");
+    BOXAssert(self.path, @"The path parameter for BOXMetadataUpdateTask cannot be nil");
+    BOXAssert(!(self.operation == BOXMetadataUpdateADD || self.operation == BOXMetadataUpdateREPLACE) || self.value, @"The value parameter cannot be nil when using BOXMetadataUpdateADD or BOXMetadataUpdateREPLACE.");
+}
+
+- (NSString *)BOXMetadataUpdateOpsToString:(BOXMetadataUpdateOps)operation
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadataUpdateTask.m]: Change to BOXMetadataUpdateOperationToString.

File: BoxContentSDK/BoxContentSDK/Requests/BOXMetadataCreateRequest.h [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+ * templates @see BOXMetadataTemplateRequest.
+ */
+@property (nonatomic, readwrite, strong) NSString *template;
+
+/**
+ * The information to populate the initial metadata's template with.
+ *
+ * **NOTE** Must only contain instances of @see BOXMetadataTask.
+ */
+@property (nonatomic, readwrite, strong) NSArray *info;
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Requests/BOXMetadataCreateRequest.h]: Rename info for clarification.

File: BoxContentSDK/BoxContentSDK/Requests/BOXMetadataCreateRequest.m [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+- (void)performRequestWithCompletion:(BOXMetadataBlock)completionBlock
+{
+    BOOL isMainThread = [NSThread isMainThread];
+    BOXAPIJSONOperation *metadataOperation = (BOXAPIJSONOperation *)self.operation;
+    
+    if (completionBlock) {
+        metadataOperation.success = ^(NSURLRequest *request, NSHTTPURLResponse *response, NSDictionary *JSONDictionary) {
+            NSLog(@"SUCCESS!");
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Requests/BOXMetadataCreateRequest.m]: Take out logging.

File: BoxContentSDK/BoxContentSDK/Requests/BOXMetadataCreateRequest.m [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+            } onMainThread:isMainThread];
+        };
+    }
+    
+    [self performRequest];
+}
+
+- (void)setInfo:(NSArray *)info
+{
+    for (NSInteger i = 0; i < info.count; ++i) {
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Requests/BOXMetadataCreateRequest.m]: Use fast enumeration.

File: BoxContentSDK/BoxContentSDK/Requests/BOXMetadataRequest.m [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    NSDictionary *queryParameters = nil;
+    
+    BOXAPIJSONOperation *JSONOperation = [self JSONOperationWithURL:URL
+                                                         HTTPMethod:BOXAPIHTTPMethodGET
+                                              queryStringParameters:queryParameters
+                                                     bodyDictionary:nil
+                                                   JSONSuccessBlock:nil
+                                                       failureBlock:nil];
+    
+    // Ask Tom about this.
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Requests/BOXMetadataRequest.m]: Remove.

File: BoxContentSDK/BoxContentSDK/Requests/BOXMetadataUpdateRequest.m [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+    
+    NSMutableArray *bodyArray = [[NSMutableArray alloc]init];
+    for (BOXMetadataUpdateTask *task in self.updateInfo) {
+        NSString *instruction = [task BOXMetadataUpdateOpsToString:task.operation];
+        NSString *path = task.path;
+        NSString *value = task.value;
+        NSMutableDictionary *taskDict = [[NSMutableDictionary alloc]init];
+        
+        if (instruction) {
+            [taskDict setObject:instruction forKey:@"op"];
```
Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Requests/BOXMetadataUpdateRequest.m]: Keys should be constants.

File: BoxContentSDK/BoxContentSDK/Requests/BOXMetadataRequest.m (hunk 16 above)
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Requests/BOXMetadataRequest.m]: remove

File: BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h [hunk 18]
```diff
@@ -30,6 +30,11 @@ extern NSString *const BOXAPIResourceCollections;
 extern NSString *const BOXAPIResourceEvents;
 extern NSString *const BOXAPIResourceCollaborations;
 extern NSString *const BOXAPIResourceSearch;
+extern NSString *const BOXAPIResourceMetadataTemplates;
+
+// API Metadata Scope
+extern NSString *const BOXAPIScopeEnterprise;
+extern NSString *const BOXAPIScopeGlobal;
```
Comment (lopopolo) [inline BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h]: I'd call these TemplateScopes lest they be confused with OAuth2 scopes.

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+ * Generate a request to retrieve metadata information about a file.
+ *
+ * **NOTE** Defaults the scope property to @see BOXAPIScopeEnterprise.
+ *
+ * @param fileID File ID.
+ * @param template Template for metadata in file.
+ *
+ * @return A request that can be customized and then executed.
+ */
+- (BOXMetadataRequest *)metadataInfoRequestWithFileID:(NSString *)fileID template:(NSString *)template;
```
Comment (lopopolo) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h:36]: this API does not exist. /files/id/metadata/properties is a one off and is an alias for scope=global template=properties.

I would rather it not be included and force users to be explicit.

Comment (lopopolo) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h:36]: oh, i see. You're being clever and anticipating that users will be requesting enterprise scoped metadata more often?

Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Metadata.h:36]: Yes. I figured that getting metadata based on custom information is more useful than the global information so having a convenience method would be nice. Realistically if they are looking for something specific they can always call the other method where they can specify their own scope.

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Search.h [hunk 20]
```diff
@@ -24,4 +24,17 @@
...
                                      inRange:(NSRange)range;
 
+/**
+ * Generate a request to search for files and folders in a user's account using metadata tags. You can customize
+ * the search by setting properties of the BOXSearchRequest before executing it.
+ *
+ * @param templateKey The template to search for metadata tags in
+ * @param scope The scope to search for metadata templates in
+ * @param filters The metadata tags to filter for
+ * @param range An offset (NSRange location) and limit (NSRange limit). The maximum limit allowed is 200, default is 30.
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+Search.h:34]: inRange to be consistent ?

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadataTemplateField.m [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+//
+
+#import "BOXMetadataTemplateField.h"
+
+@implementation BOXMetadataTemplateField
+
+- (instancetype)initWithJSON:(NSDictionary *)JSONData
+{
+    if (self = [super initWithJSON:JSONData]) {
+        self.modelID = [NSJSONSerialization box_ensureObjectForKey:@"key"
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadataTemplateField.m]: constants

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadataTemplate.m [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+#import "BOXMetadataTemplate.h"
+#import "BoxISO8601DateFormatter.h"
+
+@implementation BOXMetadataTemplate
+
+- (instancetype)initWithJSON:(NSDictionary *)JSONData
+{
+    if (self = [super initWithJSON:JSONData]) {
+        self.modelID = [NSJSONSerialization box_ensureObjectForKey:@"templateKey"
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadataTemplate.m]: constants

File: BoxContentSDK/BoxContentSDK/Models/BOXMetadataUpdateTask.m [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+- (void)validate
+{
+    BOXAssert([self BOXMetadataUpdateOperationToString:self.operation], @"The operation parameter for BOXMetadataUpdateTask cannot be nil. Please make sure to use an ENUM within BOXMetadataUpdateOperation.");
+    BOXAssert(self.path, @"The path parameter for BOXMetadataUpdateTask cannot be nil");
+    BOXAssert(!(self.operation == BOXMetadataUpdateADD || self.operation == BOXMetadataUpdateREPLACE) || self.value, @"The value parameter cannot be nil when using BOXMetadataUpdateADD or BOXMetadataUpdateREPLACE.");
+}
+
+- (NSString *)BOXMetadataUpdateOperationToString:(BOXMetadataUpdateOperation)operation
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Models/BOXMetadataUpdateTask.m:45]: comment in here somewhere about how these are not meant to be localized (functional, not user-facing)

File: BoxContentSDK/BoxContentSDK/Requests/BOXSearchRequest.m [hunk 24]
```diff
@@ -28,75 +34,100 @@ - (instancetype)initWithSearchQuery:(NSString *)query inRange:(NSRange)range
...
-
-    // If both to and from values are not set, the API allows trailing commas (e.g. "value," or ",value")
-    if (self.createdAtFromDate || self.createdAtToDate) {
-        NSString *fromDate = [self.createdAtFromDate box_ISO8601String];
-        NSString *toDate = [self.createdAtToDate box_ISO8601String];
-        queryParameters[BOXAPIParameterKeyCreatedAtRange] =
+    queryParameters[queryKey] = self.query;
+
+    if (!isMetadataSearch) {
+        if (self.requestAllItemFields) {
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Requests/BOXSearchRequest.m]: aren't these applicable to the metadata search too? I'd think request all item fields would be...

**Comment by boxtcarpel** [comment]
Some comments.

+2 LGTM

:radio_button:

**Comment by boxtcarpel** [comment]
+2
:up:

**Pull request merged**